### DISCUSSION TITLE
Purpose of "mux criteria" of the C++ example?

### DISCUSSION THREAD

**Discussion by mf01** [xR3b0rn/dbcppp#87] [Q&A] [answered]
What is the purpose of the "mux criteria"? The C++ example denotes the following criteria:
```
const dbcppp::ISignal* mux_sig = msg->MuxSignal();
if (sig.MultiplexerIndicator() != dbcppp::ISignal::EMultiplexer::MuxValue ||
    (mux_sig && mux_sig->Decode(frame.data) == sig.MultiplexerSwitchValue()))
```
After investigating the code, I did not understand the purpose of this criteria. Furthermore, when I call `sig.Decode(frame.data)` to a (correct sized) buffer. I got the expected data (in my examinated cases).

The C example has no obvious code therefore.

**Comment by xR3b0rn** [comment, marked as answer]
In addition to normal signals, DBC also offers an opportunity to specify a multiplexer signal and multiplexed signals in messages. If a multiplexer signal is specified in a message, the multiplexer signal determines which multiplexed signals are in the frame.

E.g. from `tests/test_files/dbc/multiplex.dbc`:
```
BO_ 2148676694 Message1: 8 Vector__XXX
 SG_ BIT_K m24 : 28|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_D m24 : 32|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_B m24 : 33|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_F m24 : 39|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_H m24 : 38|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_E m24 : 29|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_A m24 : 26|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_L m8 : 24|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_G m8 : 23|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_C m8 : 19|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ BIT_J m8 : 18|1@1+ (1,0) [0|0] "" Vector__XXX
 SG_ Multiplexor M : 2|6@1+ (1,0) [0|0] "" Vector__XXX
```
The message `Message1` specifies the signal `Multiplexor` (`M`) as the multiplexer signal, the other signals are multiplexed (`m<number>`) . If the multiplexer signal value equals 24, the frame contains the signals specified with `m24`,  analogue for `m8`.

So to have a generic algorithm which decodes signals from an arbritary DBC, you first have to check, if the signal is a multiplexed signal. And if so, you have to compare the value from the DBC against the multiplexer value from the data frame you've received.

**Comment by mf01** [comment]
OK, now I understand this. Thank you-

**Comment by MichaelBMiner** [comment]
This is great info, but how does one get the mux value using C? I have attempted to do this but do not properly parse the mux value. 

```
BO_ 2566544894 PGN_64093_EVSE_1_Power_Max_Schedule: 8 EVCC
 SG_ MUX M : 0|4@1+ (1,0) [0|0] "" EV
 SG_ EVSE1_Schedule_Index m0 : 4|8@1+ (1,0) [0|250] "count" EV
 SG_ EVSE1_Max_Schedule_Index m0 : 12|8@1+ (1,0) [0|250] "count" EV
 SG_ EVSE1_Schedule_Id m0 : 20|16@1+ (1,0) [0|65535] "count" EV
 SG_ EVSE1_PMax_Schedule_Id m0 : 36|16@1+ (1,0) [0|65535] "count" EV
 SG_ EVSE1_Entry_Power_Max m1 : 4|32@1- (1,-2000000000) [-2000000000|2211081215] "W" EV
 SG_ EVSE1_Entry_Start_Time m2 : 4|32@1+ (1,0) [0|4211081215] "s" EV
 SG_ EVSE1_Entry_Duration m3 : 4|32@1+ (1,0) [0|4211081215] "s" EV
```

**Comment by xR3b0rn** [comment]
You can use this function: `dbcppp_SignalMultiplexerSwitchValue`

**Reply by MichaelBMiner** [reply]
```
const dbcppp_Signal* signal = dbcppp_MessageSignals_Get(msg, j);
dbcppp_ESignalMultiplexer signal_mux_type = dbcppp_SignalMultiplexerIndicator(signal);
uint64_t signal_mux_value = dbcppp_SignalMultiplexerSwitchValue(signal);
```

No matter the can data, this always returns 0.
